File: typescript-test-playwright/tests/ui/sauce-demo/sauce-demo.test.ts
import { expect, Page, test } from '@playwright/test';
import { roles } from '@config/config';
import { CartItems, HomePage, InventoryItem, InventoryPage, ShoppingCartPage } from '@page/sauce-demo';

test.describe('Swag Labs demo app', () => {
  let page: Page;
  let homePage: HomePage;
  let inventoryPage: InventoryPage;
  let shoppingCartPage: ShoppingCartPage;

  test.beforeAll(async ({ browser }) => {
    page = await browser.newPage();
    homePage = new HomePage(page);
    inventoryPage = new InventoryPage(page);
    shoppingCartPage = new ShoppingCartPage(page);

    await homePage.open();
    await homePage.login(roles.sauceDemoRole);
  });

  test.afterAll(async () => {
    await inventoryPage.header.logout();
  });

  test('Add item to shopping cart', async () => {
    const itemToAdd = 'Backpack';

    const item = await inventoryPage.getAllItems().then((items: InventoryItem[]) => {
      return items.find((item) => {
        return item.name.includes(itemToAdd);
      });
    });
    await item?.addToCart();
    await inventoryPage.header.openShoppingCart();
    await shoppingCartPage.getCartItems().then((items: CartItems[]) => {
      expect(
        items.find((item) => item.name.includes(itemToAdd)),
        `Expected to find ${itemToAdd} from cart items`,
      ).toBeTruthy;
    });
  });
});

File: python-test/test/ui/sauce_demo/sauce_demo_test.py
import pytest
from base_selenium_test import BrowserTestCase
from pages.sauce_demo.home_page import HomePage
from pages.sauce_demo.inventory_page import InventoryPage
from pages.sauce_demo.shopping_cart_page import ShoppingCartPage

@pytest.mark.saucedemo
class ExampleTestCase(BrowserTestCase):

    def setUp(self):
        self.home_page = HomePage(self.browser)
        self.inventory = InventoryPage(self.browser)
        self.shopping_cart = ShoppingCartPage(self.browser)

    def test_this(self):
        self.home_page.open()
        self.home_page.login('standard_user', 'secret_sauce')

        inventory_items = self.inventory.get_all_items()
        [item for item in inventory_items if 'Backpack' in item.name][0].add_to_cart()

        self.inventory.header.open_shopping_cart()
        shopping_cart_items = self.shopping_cart.get_cart_items()

        assert len(shopping_cart_items) == 1
        assert 'Backpack' in shopping_cart_items[0].name

        self.shopping_cart.header.logout()
